rust/primary/src/core.rs

```diff
@@ -0,0 +1,411 @@
+    async fn process_certificate(&mut self, certificate: Certificate) -> DagResult<()> {
+        debug!("Processing {:?}", certificate);
+
+        // Process the header embedded in the certificate if we haven't already voted for it (if we already
+        // voted, it means we already processed it). Since this header got certified, we are sure that all
+        // the data it refers to (ie. its payload and its parents) are available. We can thus continue the
+        // processing of the certificate even if we don't have them in store right now.
+        if !self
+            .processing
+            .get(&certificate.header.round)
+            .map_or_else(|| false, |x| x.contains(&certificate.header.id))
+        {
+            // This function may still throw an error if the storage fails.
+            self.process_header(&certificate.header).await?;
+        }
+
+        // Ensure we have all the ancestors of this certificate yet. If we don't, the synchronizer will gather
+        // them and trigger re-processing of this certificate.
+        if !self.synchronizer.deliver_certificate(&certificate).await? {
+            debug!(
+                "Processing of {:?} suspended: missing ancestors",
+                certificate
+            );
+            return Ok(());
+        }
+
+        // Store the certificate.
+        let bytes = bincode::serialize(&certificate).expect("Failed to serialize certificate");
+        self.store.write(certificate.digest().to_vec(), bytes).await;
+
+        // Check if we have enough certificates to enter a new dag round and propose a header.
+        if let Some(parents) = self
+            .certificates_aggregators
+            .entry(certificate.round())
+            .or_insert_with(|| Box::new(CertificatesAggregator::new()))
+            .append(certificate.clone(), &self.committee)?
+        {
+            // Send it to the `Proposer`.
+            self.tx_proposer
+                .send((parents, certificate.round()))
+                .await
+                .expect("Failed to send certificate");
+        }
+
+        // Send it to the consensus layer.
+        let id = certificate.header.id.clone();
+        if let Err(e) = self.tx_consensus.send(certificate).await {
+            warn!(
+                "Failed to deliver certificate {} to the consensus: {}",
+                id, e
+            );
+        }
+        Ok(())
+    }
+
+    fn sanitize_header(&mut self, header: &Header) -> DagResult<()> {
+        ensure!(
+            self.gc_round <= header.round,
+            DagError::TooOld(header.id.clone(), header.round)
+        );
+
+        // Verify the header's signature.
+        header.verify(&self.committee)?;
+
+        // TODO [issue #3]: Prevent bad nodes from sending junk headers with high round numbers.
+
+        Ok(())
+    }
+
+    fn sanitize_vote(&mut self, vote: &Vote) -> DagResult<()> {
+        ensure!(
+            self.current_header.round <= vote.round,
```

Comment:
why not just equal?

Reply:
it is actually an equal. But we want to throw different errors for when we receive a vote for a past round (ie. not one is bad, the vote simply arrived late) VS when we received a vote for a block we didn'tcreate (ie. someone is bad).:

https://github.com/facebookresearch/narwhal/blob/1ff0d968e8a01a6c97015497baf5fb996c38fa21/rust/primary/src/core.rs#L326-L330